Parameterized scenario test

lechlm opened the pull request:
I have created a small parameterized test, which takes two files as input and checks, if the the correct output was produced.

The expectedOutput file contains all board states sequentially, whereas the input file contains at the first line the board size, then the config file, and afterwards all user inputs.

With this generic test it's quite easy to setup different scenarios for testing various aspects.

The only thing I'm not sure about is, if the make file must be somehow updated? So far I only tested it in my IDE, so it might be good to clarify this before merging.

lechlm committed:
Created scenario test.

Wooza commented:
Nice implementation!
I have previously added two scenario tests to `BasicTest.java`, however, I think that your approach is better suited for modelling scenarios.
Can you expand this PR to delete my tests (`testLossWithInitialMineHit` and `testWinWithInitialMineAvoided`), since they are superseded by your approach?

alessiogambi reviewed (commented):
I need some time to look carefully at this. Personally, I do not like parametrized tests because it is harder to understand what's going on (or wrong) when the number of tests files increases (which file failed?). Having very generic oracles that simply check the entire file does not match, is a good start, but a proper matcher should be developed as well that exactly pin/points the problem. Basically, when a test fails the assertion should be as specific as possible.
Finally, creating the input/output files is a bit of a P.I.T.A.

alessiogambi commented on test_scenarios/1/lostGame-input.txt:
```diff
@@ -0,0 +1,8 @@
+4
```
Why there's only a number here? The boards can have a rectangular shape, so I guess you need to add also the number of columns/row? If then you specify the board (with `.` and `*`) why do you need to input also the size of the board?

The format of the input file is a bit messy as you put in the same file different into (board, actions). This prevents reusing the same board (file) to test different input sequences.

lechlm commented:
Regarding the point of determining the failed test files, I used the file name without the suffix as an indicator because in my opinion this makes it quite clear, which scenario has failed. 
Regarding the used assertions, I fully agree, a matcher should be definitely developed to exactly determine, which part of the scenario has failed. I will do this in the upcoming days and add it to this PR.
Furthermore, I know that creating the input/output files is not so nice, but I thought for testing various (also more complicated) game setups, this is a reasonable approach, at least for the first assignment, where everything is in one method.

alessiogambi commented:
Don't get me wrong. This approach is indeed useful and in previous instances of Programming Styles has been adopted always. The part I did not like, is that once this is in place then the students use only and only this method for testing their application, which seems a bit overkilling in many cases. For example, imagine you want to test that "flagging twice one square results in the same state the square had originally". Using a scenario-based test, one has to create all the input/output files, and likely consider the entire GUI in the assertions. Capturing this with an "uber" matcher/assertion is also not advisable because it is really complex.

Another aspect that is not good for testing, in general, is that those scenario tests might check too much, as they check basically every aspect of the code (states, error messages, board evolutions, etc.).

lechlm commented:
Okay, I think, I got the point. Thank you for explaining it. When not further pursuing the idea of the scenario test, an open question in my opinion is then the point, how intermediate results can be checked "efficiently" (e.g. your example with flagging twice one square) because we only receive the text output from the program. But I have to admit, I haven't thought about that too much till now because I first went with the other idea.

alessiogambi commented:
Let's first finish this pull request and merge it, so we have a starting point for thinking about the other issues. Is this code ready to be merged?

lechlm committed:
Removed two basic tests, which are superseded by the scenario test, and added them as scenario test files.

lechlm commented:
I have removed the two basic tests testLossWithInitialMineHit and testWinWithInitialMineAvoided, as suggested by @Wooza and added them as test files. Apart from that, the code is in my opinion ready to be merged.

alessiogambi reviewed (approved):
I still cannot understand why you need to pass both the size of the board and its content. This is redundant and prevents one to define rectangular boards. I guess we will address this in a separate issue

alessiogambi committed:
Merge branch 'main' into parameterized-scenario-test
